src/email/outbound.js

```diff
@@ -0,0 +1,326 @@
+    const smtpPort = options.smtpPort || 25;
+    const smtpHost = options.smtpHost;
+    const rejectUnauthorized = options.rejectUnauthorized;
+    const startTLS = options.startTLS || false;
+    const tls = options.tls || {};
+
+    // group recipients by domain(to limit the number of connections per 'To' domain).
+    function groupRecipients(recipients) {
+        const groups = {};
+        for (const recipient of recipients) {
+            const host = ParseEmailAddress.parseOneAddress(recipient).domain;
+            (groups[host] || (groups[host] = [])).push(recipient);
+        }
+        return groups;
+    }
+
+    function getAddresses(addresses) {
+        const results = [];
+        if (!Array.isArray(addresses)) {
+            addresses = addresses.split(',');
+        }
+        const addressesLength = addresses.length;
+        for (let i = 0; i < addressesLength; i++) {
+            results.push(ParseEmailAddress.parseOneAddress(addresses[i]).address);
+        }
+        return results;
+    }
+
+    /**
+     * Resolve MX records by domain.
+     * @param domain
+     * @returns {Promise<MxRecord[]|*[]>}
+     */
+    async function resolveMX(domain) {
+        let resolvedMX = [];
+        if (smtpHost !== '') {
+            resolvedMX.push({ exchange: smtpHost });
+            return resolvedMX;
+        }
+        try {
+            resolvedMX = await resolver.resolveMx(domain);
+            resolvedMX.sort(function (a, b) { return a.priority - b.priority; });
+            return resolvedMX;
+        }
+        catch (ex) {
+            throw Error(`Failed to resolve MX for ${domain}: ${ex}`);
+        }
+    }
+
+    async function sendToSMTP(domain, srcHost, from, recipients, body) {
+        const resolvedMX = await resolveMX(domain);
+        log.info("Resolved mx list:", resolvedMX);
+        let sock;
+        function tryConnect(i) {
+            if (i >= resolvedMX.length) {
+                throw Error(`Could not connect to any SMTP server for ${domain}`);
+            }
+
+            sock = createConnection(smtpPort, resolvedMX[i].exchange);
+
+            sock.on('error', function (err) {
+                log.error('Error on connectMx for: ', resolvedMX[i], err);
+                tryConnect(++i);
+            });
+
+            sock.on('connect', function () {
+                log.debug('MX connection created: ', resolvedMX[i].exchange);
+                sock.removeAllListeners('error');
+                return sock;
+            });
+        }
+        tryConnect(0);
+
+        function w(s) {
+            log.debug(`SEND ${domain}> ${s}`);
+            sock.write(s + CRLF);
+        }
+
+        sock.setEncoding('utf8');
+
+        sock.on('data', function (chunk) {
+            data += chunk;
+            parts = data.split(CRLF);
+            const partsLength = parts.length - 1;
+            for (let i = 0, len = partsLength; i < len; i++) {
+                onLine(parts[i]);
```

Comment:
Reccomendation would be to have the function definitions above the calls, I was surprised by the `onLine` call and had to find it further below. Just aids readability :)